File: medium/EncodeandDecodeStrings.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  string encode(vector<string> &strs) {
    string ans;
    for (string str : strs) {
      int size = str.size();
      string sizeStr = to_string(size) + '#' + str;
      ans += sizeStr;
    }
    return ans;
  }

  vector<string> decode(string s) {
    vector<string> ans;
    int i = 0;
    while (i < s.size()) {
      string next_word_chars_count = "";
      while (s[i] != '#') {
        next_word_chars_count += s[i];
        i++;
      }
      int word_size = stoi(next_word_chars_count);
      i++;
      string current_word = "";
      while (word_size > 0) {
        current_word += s[i];
        i++;
        word_size--;
      }
      ans.push_back(current_word);
    }
    return ans;
  }
};

int main() {}

File: ts/medium/EncodeandDecodeStrings.ts
class Solution {
	encode(strs: string[]): string {
		let encode = "";
		for (const str of strs) {
			encode += `${str.length}#` + str;
		}
		return encode;
	}

	decode(str: string): string[] {
		let decode: string[] = [];
		let i = 0;
		while (i < str.length) {
			let counter = 0;
			let numberStr = "";
			while (/[0-9]/.test(str[i]) && /[0-9]/.test(str[i])) {
				numberStr += str[i];
				i += 1;
			}
			i++;
			counter = parseInt(numberStr);
			decode.push(str.slice(i, i + counter));
			i += counter;
		}
		return decode;
	}
}
